[38690] Email design - Daily Reminders

HDinger opened the pull request:
### ToDo

- [x] The daily email should contain all unread notifications that happened since last time you received a daily reminder.
- [x] The header should include:
  - [x] Logo of Open project.
  - [x] Salutation with the name of the user.
  - [x] Information: When this email is generated (for search reasons), “X” unread notifications, in “X” Work Packages.
  - [x] A button access to the notification centre (main view with no work package selected).
- [x] The content should be divided in notification rows as in the Notification Centre (with some UI variations as card shadows).
  - [x] Aggregate all notifications for one Work Package 
  - [x] In this case all the lines of changes will be displayed (if there is 3 notifications in a single Work Package each of them will be displayed in a line).
  - [x] Show configured colours for type and status
  - [x] When the user clicks in one of the rows the Notification Centre opens with the specific activity for that work package open.
  - [x] Also update text only version to match the new styled email
- [x] A button with a link to the email settings is displayed at the end of the email.
- [x] A limit of 15 work packages will be applied to this mails. If the user has more than 15 work packages with notifications a "show all work packages" button will be displayed that will lead to the Notification Centre main view.
- [x] Check for small screens
- [x] Adapt tests

https://community.openproject.org/projects/openproject/work_packages/38690/activity

HDinger committed:
Add basic layout for showing notification rows in the daily email

HDinger committed:
Link to notification center from the email notification

HDinger committed:
Adapt text-only version of the daily email

HDinger committed:
Highlight type and status in email in the defined colors

HDinger committed:
Add header block for daily email with general information like salutation and number of notifications

HDinger committed:
Add button to change notification settings in email

HDinger committed:
Limit the number of shown WPs in the daily email

HDinger committed:
Show comment in notification row in daily email

HDinger committed:
Add logo to email header block

HDinger committed:
Stabilise email for the case that status or type don't have a defined color

HDinger committed:
Adapt test to new email layout

HDinger committed:
Differentiate between normal comments and @mentioned comments

HDinger committed:
Extract shared email parts into partials to make it easier to reuse them later

ulferts reviewed (changes requested):
I checked the resulting mail with the following clients:
* Thunderbird on mac
* Mail on mac
* mailbox.org UI using Chrome on mac
* Mail on iPhone

When reviewing I did not have access to the figma files yet, so I am using the screenshots shown in https://community.openproject.org/wp/38690 as a reference. 

Some of the deviations I find acceptable personally and could understand that judging them by an effort vs. gain equation might result in them not being fixed.

# Thunderbird on mac

* The logo is not displayed at the top:
 
![image](https://user-images.githubusercontent.com/617519/131998825-7aa449a4-5a5b-4b3b-8f81-4860a4483413.png)

From what I see in the code, that icon would always be the OpenProject icon. This might not be desired by every customer if that customer has uploaded a custom icon. Admittedly, this has not been specified.

* There are no icons in the "notification center" button and in the "Change email settings" button. This seems to be accepted as Out of scope, though.
![image](https://user-images.githubusercontent.com/617519/131999422-cf81142a-52cf-4784-8cae-c2c9d22d7865.png)
![image](https://user-images.githubusercontent.com/617519/132000007-0df28ce1-b556-4e1d-83b0-35703848e254.png)


* The 'id - project - reasons' line is not printed in bold
![image](https://user-images.githubusercontent.com/617519/131999557-93d182b2-a2ca-4065-a5bc-959c9c5e010c.png)

* No drop shadow around each notification (after having received access to figma, this can probably be dropped since there are no shadows in figma)
![image](https://user-images.githubusercontent.com/617519/131999805-87744f59-d0fd-4efb-81be-1e86ee578138.png)

* If the mail window's width is decreased, and longer lines are present in the notifications and the user then scrolls to the right, lines breaking out of the notification boxes become visible

![image](https://user-images.githubusercontent.com/617519/132000348-6fed8fd6-200c-425e-a5eb-3c9ba45829f4.png)

* The notification settings will be split up into two pages (notification settings and email reminders). The link on the "Change email settings" button currently points to the notification settings which is ok since the other does not exist yet. But either the wording or the target should be changed in the long run.

* The logo is embedded at the bottom of the mail:
![image](https://user-images.githubusercontent.com/617519/132000952-d51e694b-f6d2-435f-b5bc-785b5a6e2e83.png)
 
* The subject in the screenshot has a different format: OpenProject (or instance name?) - x unread notifications in y work packages, in z projects
 
![image](https://user-images.githubusercontent.com/617519/132001302-494ee2e7-453f-49cf-9357-4c3faae34305.png)

* The margins between boxes and at various other places (e.g. between the horizontal row and the first notification box) seem to be less than in the screenshot.

# Mail on mac

* Because of the dark mode, parts of the mail are very hard to read
![image](https://user-images.githubusercontent.com/617519/132001779-3758e5c7-db57-459e-9bf0-26493887c652.png)
![image](https://user-images.githubusercontent.com/617519/132001929-fbbf0e12-d64b-41e3-8d91-0cb0fe1ad9f5.png)
![image](https://user-images.githubusercontent.com/617519/132001947-9d9b9c55-a2d3-45b1-927a-d72a1a7b0c9f.png)



* The logo is embedded at the bottom (but differently) and missing at the top
![image](https://user-images.githubusercontent.com/617519/132001889-645b4352-fa6f-477c-a76b-0481e53971ae.png)

# mailbox.org UI using Chrome on mac

* No rounded button for the "Notification center" link (the button to "Change email settings" is rounded) 
![image](https://user-images.githubusercontent.com/617519/132002042-5abfe7ac-fc8c-4b8c-ab58-8774f3f1eb23.png)

* Increased height for the status badge
* Id written vertically
* Type written vertically
![image](https://user-images.githubusercontent.com/617519/132002338-7c547a7d-1a20-4839-8822-66b97f8f7f70.png)

# Mail on iPhone

* Inconsistent font size between notifications
* notification count pill not floating right
![image](https://user-images.githubusercontent.com/617519/132003053-6bb6a4ce-4072-434f-9265-41b5717552b6.png)

HDinger committed:
Adjust spacings and alignment in daily email

HDinger committed:
Load image via Base64 string instead of direct attachment

HDinger committed:
Change subject string (maybe changed later)

HDinger committed:
Show custom logo in the mail if available

HDinger committed:
Put all top-level information in one table cell to handle overflows better

HDinger committed:
Enforce white background even in dark mode to match the styling requirements

ulferts commented on app/helpers/mail_notification_helper.rb:
```diff
@@ -0,0 +1,74 @@
+  include ::ColorsHelper
+
+  def logo_tag(**options)
+    current_logo = CustomStyle.current.logo
```
`CustomStyle.current` might be nil resulting in a an error when sending `logo` to it.

ulferts reviewed (changes requested):
# Contents regardless of client

* The reasons for notification use different wording compared to the Figma file
e.g.
![image](https://user-images.githubusercontent.com/617519/132471874-d1221d17-45d1-4314-a593-ebcf109e442c.png)

vs 

![image](https://user-images.githubusercontent.com/617519/132472127-bd51d30f-1470-45cc-aeca-6ebb1cdd5018.png)
The Figma file seems to expect to differentiate between "Assignee" and "Accountable" as well which is out of scope for this PR

* The headings for the project are no longer specified to be bold. This seems to have been changed in the meantime.

* The subject line does not note down the number or projects: e.g. "23 unread notification(s) in 8 work package(s)" 

* The link "Change email settings" should point to the "reminder settings" (clarified only recently)

* The "status badges" in the Figma file have more padding to the left and right compared to the implementation.
Mail:
![image](https://user-images.githubusercontent.com/617519/132474896-6dfae589-917c-4f51-94e1-87be8f8e8763.png)
Figma:
![image](https://user-images.githubusercontent.com/617519/132474947-61883ddb-5034-41d0-925b-d475c429891c.png)

# Thunderbird on mac 

* The details sometimes break out of the boxes when decreasing the width of the mail window

![image](https://user-images.githubusercontent.com/617519/132471590-e0dfd0c0-1f63-4698-b2b7-49f25c6f68a1.png)



# Mail on mac 

Same problems as in Thunderbird

# Mailbox.org UI in Chrome on Mac

Same problems as in Thunderbird

# Mail on iOS

* Different font sizes between notification rows

![image](https://user-images.githubusercontent.com/617519/132480937-53a30c4d-ed41-4b35-a6b4-5956ca912941.png)

This might be something iOS does in order to increase readability but that is just guessing.

* Personal preference: I don't like the way long headings caused by project names are displayed on mobile devices. To me, they should be displayed in their own row but that would require media queries and I don't assume them to be supported.

![image](https://user-images.githubusercontent.com/617519/132481651-026ab320-795a-4810-9c50-cdf3f6c2ac57.png)

# GMX in Chrome on Mac

* rounded corners are not supported at all. I think the outcome is acceptable, though.

![image](https://user-images.githubusercontent.com/617519/132482109-75760e86-f28a-41e1-ab7c-13c24bbb38e9.png)

# Gmail in Chrome on Mac

All fine. They seem to have implemented a min width.

HDinger committed:
* Show timestamp instead of relative time ago in the daily mail
* Change url of settings button

HDinger committed:
increase font-sizes to a minimum of 14px to avoid iOS doing its own "optimisations" and increasing the fonts randomly (taken from https://monsterspost.com/render-emails-perfectly-different-email-clients/)

HDinger committed:
* Change email subject and summary text to include how many direct mentioned there have been
* Move logo to the right side

HDinger commented:
I have addressed most of the issues. There are of course still some limitations we have to live with (e.g. rounded corners are not supported by all clients). The concrete text for the subject and the header part is currently still discussed so I'd leave it like this for now and change it later separately if needed.

ulferts reviewed (approved):
Looking good design-wise now, @HDinger 🥇 🎈 👍

HDinger committed:
Change texts to handle singular vs plural better

HDinger commented:
Thanks a lot for your ongoing reviews @ulferts ! 👏 I'll merge this then once I got the tests green.
